src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/RestMaturedBlockRequester.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Stratis.Bitcoin.Utilities;
+using Stratis.FederatedPeg.Features.FederationGateway.Controllers;
+using Stratis.FederatedPeg.Features.FederationGateway.Interfaces;
+using Stratis.FederatedPeg.Features.FederationGateway.Models;
+using Stratis.FederatedPeg.Features.FederationGateway.SourceChain;
+
+namespace Stratis.FederatedPeg.Features.FederationGateway.TargetChain
+{
+    public class RestMaturedBlockRequester : RestSenderBase, IMaturedBlocksRequester, IDisposable
+    {
+        public const int MaxBlocksToCatchup = 1000;
+        public static TimeSpan CatchUpInterval = TimeSpans.TenSeconds;
+
+        private IAsyncLoop asyncLoop;
+        private IAsyncLoopFactory asyncLoopFactory;
+        private ICrossChainTransferStore crossChainTransferStore;
+        private INodeLifetime nodeLifetime;
+        private int maxDepositHeight;
+
+        public RestMaturedBlockRequester(
+            ILoggerFactory loggerFactory,
+            IFederationGatewaySettings settings,
+            IHttpClientFactory httpClientFactory,
+            IAsyncLoopFactory asyncLoopFactory,
+            ICrossChainTransferStore crossChainTransferStore,
+            INodeLifetime nodeLifetime)
+            : base(loggerFactory, settings, httpClientFactory)
+        {
+            this.asyncLoopFactory = asyncLoopFactory;
+            this.crossChainTransferStore = crossChainTransferStore;
+            this.nodeLifetime = nodeLifetime;
+        }
+
+        /// <inheritdoc />
+        public void SetTip(int tipHeight)
+        {
+            this.maxDepositHeight = tipHeight;
+        }
+
+        /// <inheritdoc />
+        public void SetLastReceived(int lastReceived)
+        {
+            if (lastReceived > this.maxDepositHeight)
+            {
+                this.maxDepositHeight = lastReceived;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Start()
```

Comment:
rather than having a big loop with sleeps (pull model) could we try to react to events of blocks appearing ?
Maybe MaturedBlockReceiver.MaturedBlockDepositStream can be observed and the only source of changes to this.maxDepositHeight ?
In my opinion it would make it easier to reason about this component, here for instance I am not sure what can happen to this.maxDepositHeight when different parts of the code try to change it at the same time...

Reply:
`The push model is inefficient as it sends one block at a time`... I don't think this is a limitation of the push model itself, what prevents you from batching the blocks you receive or anything similar ?
this is an example
http://introtorx.com/Content/v1.0.10621.0/13_TimeShiftedSequences.html
I think the result is a lot easier to read generally

Reply:
Why would we want to buffer data until conditions are ideal for consumption (sending will be faster than storage). This imposes unnecessary memory consumption when we can instead have the target drive the pace?

Reply:
If you are going to request the blocks anyway (by some method taking a numBlocks argument) then you might as well return the result at that time. Why take a longer route and first have a second remote call (able to send only one block at a time currently) to send them to the receiver so the receiver can then pop them in a stream? The stream may be easy to test but why add something just because its easy to test if it's really not needed at all? Too many extra steps which makes it unlikely that it will match the current performance of around 1 minute for over 654,000 blocks.

Reply:
Cf my comment in private